src/main/java/org/broadinstitute/hellbender/engine/ReadInput.java

```diff
@@ -0,0 +1,170 @@
+ *
+ *     --argument_name logical_name:read_file
+ *
+ * These logical names can then be retrieved by the tool at runtime via {@link #getName}
+ *
+ * Furthermore, a list of comma-separated key=value pairs may be provided as follows:
+ *
+ *     --argument_name logical_name,key1=value1,key2=value2:read_file
+ *
+ * the string value provided for a given key can be retrieved via {@link #getAttribute(String)}. Keys must be unique.
+ *
+ */
+public final class ReadInput {
+
+    /**
+     * Logical name for this source of Features optionally provided by the user on the command line
+     * using the --argument_name logical_name:read_file syntax. Defaults to the absolute path of
+     * the underlying file if no logical name is specified
+     */
+    private final String name;
+
+    private final Map<String, String> kevValueMap;
+
+    /**
+     * File containing Reads as specified by the user on the command line
+     */
+    private final File readFile;
+
+    /**
+     * Construct a ReadInput from a String argument value either of the form "logical_name:read_file"
+     * or simply "read_file".
+     *
+     * Only meant to be called by the argument parsing system, and therefore marked as package-visible --
+     * ReadInputs constructed some other way will not be recognized by the engine.
+     *
+     * Note: cannot delegate to another constructor because Java only allows a call to "this" on the first line of a constructor.
+     *
+     * @param rawArgumentValue String of the form "logical_name:read_file" or "read_file"
+     */
+    ReadInput(final String rawArgumentValue) {
+        Utils.nonNull(rawArgumentValue, "rawArgumentValue");
+        final ParsedArgument parsedArgument = ParsedArgument.of(rawArgumentValue);
+
+        name = parsedArgument.getName();
+        kevValueMap = parsedArgument.keyValueMap();
+        readFile = parsedArgument.getFile();
+    }
+
+    /**
+     * Construct a ReadInput from raw components: name, key value pairs and the file.
+     *
+     * This constructor is meant to be called by the engine and test classes --
+     * ReadInputs constructed some other way will not be recognized by the engine.
+     */
+    @VisibleForTesting
+    public ReadInput(final String name, final Map<String, String> kevValueMap, final File readFile) {
+        Utils.nonNull(name, "name");
+        Utils.nonNull(kevValueMap, "kevValueMap");
+        Utils.nonNull(readFile, "readFile");
+        this.name = name;
+        this.kevValueMap = Collections.unmodifiableMap(new LinkedHashMap<>(kevValueMap));   //make a unmodifiable copy
+        this.readFile = readFile;
+    }
+
+    /**
+     * Gets the value for the given key associated with this Reads source or {@code null}
+     * if no value is associated with a given key.
+     * @throws IllegalArgumentException if the key is {@code null}.
+     */
+    public String getAttribute(final String key) {
+        Utils.nonNull(key);
+        return kevValueMap.get(key);
```

Comment:
FeatureInput stays. But I deleted `ReadInput` - it was a wrapper for 1 pointer and forwarded all calls to that object. So, it was just useless and i deleted it. Now `TaggedInputFileArgument` can be used to represent any file argument that may have a symbolic name and key-value pairs.